### PR TITLE
Rewrite Duopull as Cloud Function

### PR DESCRIPTION
**Pull request by ajvb**
* Add support for V2 Auth logs endpoint
* Move to using PubSub to trigger, writing events to Stackdriver, and using CloudKMS encrypted keys
* Add cmd/main.go for local testing

### CODE REVIEW

File: duopull/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Cloud Function for Duo Log Collection
+
+This is a Cloud function that pulls administration/authentication/telephony logs
+from the Duo API and pushes them into a logging pipeline via a Kinesis stream.
```
Comment (ameihm0912) [inline duopull/README.md]: There are a bunch of references to Kinesis in the documentation and also in comments throughout the code; we should probably change these to reflect Stackdriver use?

File: duopull/duopull.go [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+	PROJECT_ID string
+	KEYNAME    string
+
+	cfg config
+)
+
+func init() {
+	mozlogrus.Enable("duopull")
+	PROJECT_ID = os.Getenv("GCP_PROJECT")
+	KEYNAME = os.Getenv("KMS_KEYNAME")
```
Comment (ameihm0912) [inline duopull/duopull.go:60]: Are `GCP_PROJECT` and `KMS_KEYNAME` required variables? If so we should probably document that with the other required variables in the README file, and potentially check for these in `validate`?

File: duopull/duopull.go [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+// specifically included so we can inspect the value to confirm the request was
+// successful.
+//
+// See also https://duo.com/docs/adminapi#api-details
+type logRecords struct {
+	Stat     string        `json:"stat"`
+	Response []interface{} `json:"response"`
+}
+
+type logRecordsV2 struct {
```
Comment (ameihm0912) [inline duopull/duopull.go]: Do we need both `logRecords` and `logRecordsV2`? Also the version 2 types are missing documentation.

File: duopull/duopull.go [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+	Stat     string               `json:"stat"`
+	Response logRecordsV2Response `json:"response"`
+}
+
+type logRecordsV2Response struct {
+	Authlogs []interface{} `json:"authlogs"`
+}
+
+// emitEvent is an event which will be submitted to Kinesis
```
Comment (ameihm0912) [inline duopull/duopull.go]: Additional Kinesis reference here.

File: duopull/duopull.go [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+	if err != nil {
+		return 0, err
+	}
+	if p.Timestamp == 0 {
+		return 0, fmt.Errorf("event had no timestamp")
+	}
+	return p.Timestamp, nil
+}
+
+// emitter stores all collected events for distribution to Kinesis
```
Comment (ameihm0912) [inline duopull/duopull.go]: Additional Kinesis reference here.

File: duopull/duopull.go [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+type emitter struct {
+	events []emitEvent
+}
+
+// emit batches collected events to the configured Kinesis stream
```
Comment (ameihm0912) [inline duopull/duopull.go]: Additional Kinesis reference here.

File: duopull/duopull.go [hunk 7]
```diff
@@ -0,0 +1,698 @@
...
+		}
+		out, err := toMozLog(cv)
+		if err != nil {
+			return err
+		}
+		e, err := json.Marshal(out)
+		if err != nil {
+			return err
+		}
+		logger.Log(stackdriver.Entry{Payload: json.RawMessage(e)})
```
Comment (ameihm0912) [inline duopull/duopull.go:307]: Unrelated to this PR, but we may need to adjust the pipeline parser for this (just adding it here as a note).

File: duopull/duopull.go [hunk 8]
```diff
@@ -0,0 +1,698 @@
...
+
+// minTime stores state related to the mintime parameter for the Duo API logging
+// endpoints
+type minTime struct {
+	Administrator  int `json:"administrator"`  // mintime for administrator logs
+	Authentication int `json:"authentication"` // mintime for authentication logs
+	Telephony      int `json:"telephony"`      // mintime for telephony logs
+}
+
+// load pulls mintime state information from the S3 bucket
```
Comment (ameihm0912) [inline duopull/duopull.go]: S3 reference here.

File: duopull/duopull.go [hunk 9]
```diff
@@ -0,0 +1,698 @@
...
+	ret := make([]emitEvent, 0)
+	for _, v := range l.Response {
+		ret = append(ret, emitEvent{Path: path, Event: v})
+	}
+
+	return ret, nil
+}
+
+// logRequestV2
+func logRequestV2(d *duoInterface, mintime int, path string) ([]emitEvent, error) {
```
Comment (ameihm0912) [inline duopull/duopull.go]: It looks like there is a lot of duplicate code between this and `logRequest`, do we need both of them? No problem if so, just wondering if we still need to support the other version.

**Review by ameihm0912 — CHANGES_REQUESTED** [review]

File: duopull/duopull.go (hunk 9 above)
Comment (ajvb) [reply duopull/duopull.go]: Yeah, we do as there is no v2 for the admin and telephony logs.

File: duopull/duopull.go (hunk 2 above)
Comment (ajvb) [reply duopull/duopull.go:60]: Yeah, they are required in certain settings. Added to docs and added checks.

**Review by ameihm0912 — APPROVED** [review]

**Pull request merged**